timefold-solver/src/test/java/com/baeldung/timefoldsolver/ShiftScheduleConstraintProviderTest.java

```diff
@@ -0,0 +1,50 @@
+package com.baeldung.timefoldsolver;
+
+import java.time.LocalDate;
+import java.util.Set;
+
+import ai.timefold.solver.test.api.score.stream.ConstraintVerifier;
+import org.junit.jupiter.api.Test;
+
+class ShiftScheduleConstraintProviderTest {
+
+    private static final LocalDate MONDAY = LocalDate.of(2030, 4, 1);
+    private static final LocalDate TUESDAY = LocalDate.of(2030, 4, 2);
+
+    ConstraintVerifier<ShiftScheduleConstraintProvider, ShiftSchedule> constraintVerifier = ConstraintVerifier.build(
+            new ShiftScheduleConstraintProvider(), ShiftSchedule.class, Shift.class);
+
+    @Test
+    void atMostOneShiftPerDay() {
```

Comment:
We have a convention to use the given_when_then naming convention for test names, I think you can find some examples in the style guide

Reply:
No problem:

whenTwoShiftsOnOneDay_thenPenalize
whenTwoShiftsOnDifferentDays_thenDoNotPenalize
whenEmployeeLacksRequiredSkill_thenPenalize
whenEmployeeHasRequiredSkill_thenDoNotPenalize

Reply:
My recommendation:

whenTwoShiftsOnOneDay_thenPenalize
-> givenTwoShiftsOnOneDay_whenApplyingAtMostOneShiftPerDayConstraint_thenPenalizeTwice

whenTwoShiftsOnDifferentDays_thenDoNotPenalize
-> givenTwoShiftsOnDifferentDays_whenApplyingAtMostOneShiftPerDayConstraint_thenDoNotPenalize

whenEmployeeLacksRequiredSkill_thenPenalize
-> givenEmployeeLacksRequiredSkill_whenApplyingRequiredSkillConstraint_thenPenalize

whenEmployeeHasRequiredSkill_thenDoNotPenalize
-> givenEmployeeHasRequiredSkill_whenApplyingRequiredSkillConstraint_thenDoNotPenalize

Reply:
@ge0ffrey ping here, the current names are not in BDD style, you currently have bits of the given merged into the when

Reply:
Changing
  givenTwoShiftsOnOneDay_whenApplyingAtMostOneShiftPerDayConstraint_thenPenalizeTwice
to
   givenTwoShiftsOnOneDay_whenApplyingAtMostOneShiftPerDayConstraint_thenPenalize

The penalize twice shouldn't be made into the spec. It penalizes, so that's fine. It doesn't matter by how much.

The current impl is good enough: a simple impl that works (even in production), but the unique pair approach (which requires a bunch more code due to the @PlanningId requirement) is more perfect. Perfect is the enemy of good here, for a tutorial.

Reply:
Fixed